com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/modelinference/InferredModelAssociator.java

```diff
@@ -215,6 +233,69 @@ public void discardDerivedState(final DerivedStateAwareResource resource) {
     }
     resourcesContentsList.removeAll(derived);
     getSourceToInferredModelMap(resource).clear();
+    getInferredModelToSourceMap(resource).clear();
+  }
+
+  private static <K, V> Map<K, V> newLinkedHashMapWithCapacity(final int capacity) {
+    return new LinkedHashMap<K, V>(capacity) {
+      private static final long serialVersionUID = 1L;
+
+      @Override
+      public void clear() {
+        if (isEmpty()) {
+          return;
+        }
+        super.clear();
+      }
+    };
+  }
+
+  private static <K, V> void putIntoSmallSetMap(final K key, final V value, final Map<? super K, Set<V>> map, final boolean head) {
+    Set<V> set = map.get(key);
+    if (head) {
+      if (set == null) {
+        set = new LinkedListBasedSet<V>();
+        map.put(key, set);
+      }
+      set.remove(value);
+      ((LinkedListBasedSet<V>) set).elements.addFirst(value);
+    } else {
+      if (set == null) {
+        set = new LinkedListBasedSet<V>();
+        map.put(key, set);
+      }
+      set.add(value);
+    }
+  }
+
+  /**
+   * Set implementation based on {@link LinkedList}.

```

Comment:
we could also leave some short comment regarding the origin/purpose of this class